File: go/events.go
package spec

import (
	"encoding/json"
	"fmt"
)

// Streaming event vocabulary for the universal ChatMessage spec. Mirror of
// typescript/src/events.ts and python events.py. The wire shape is a flat JSON
// object discriminated by the "event" field.

// Event names.
const (
	EventMessageStarted   = "message_started"
	EventPartAppended     = "part_appended"
	EventTokenDelta       = "token_delta"
	EventPartUpdated      = "part_updated"
	EventMessageFinalized = "message_finalized"
)

// StreamEvent is implemented by all concrete stream event types.
type StreamEvent interface {
	EventName() string
}

// MessageStartedEvent announces a new message and its initial state.
type MessageStartedEvent struct {
	Message ChatMessage
}

func (MessageStartedEvent) EventName() string { return EventMessageStarted }

func (e MessageStartedEvent) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		Event   string      `json:"event"`
		Message ChatMessage `json:"message"`
	}{Event: EventMessageStarted, Message: e.Message})
}

// PartAppendedEvent appends a content part at an index.
type PartAppendedEvent struct {
	MessageID string
	Index     int
	Part      ContentPart
}

func (PartAppendedEvent) EventName() string { return EventPartAppended }

func (e PartAppendedEvent) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		Event     string      `json:"event"`
		MessageID string      `json:"message_id"`
		Index     int         `json:"index"`
		Part      ContentPart `json:"part"`
	}{Event: EventPartAppended, MessageID: e.MessageID, Index: e.Index, Part: e.Part})
}

// TokenDeltaEvent appends streamed text to a text/reasoning part.
type TokenDeltaEvent struct {
	MessageID string
	Index     int
	Text      string
}

func (TokenDeltaEvent) EventName() string { return EventTokenDelta }

func (e TokenDeltaEvent) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		Event     string `json:"event"`
		MessageID string `json:"message_id"`
		Index     int    `json:"index"`
		Text      string `json:"text"`
	}{Event: EventTokenDelta, MessageID: e.MessageID, Index: e.Index, Text: e.Text})
}

// PartUpdatedEvent merges a patch into an existing part.
type PartUpdatedEvent struct {
	MessageID string
	Index     int
	Patch     map[string]json.RawMessage
}

func (PartUpdatedEvent) EventName() string { return EventPartUpdated }

func (e PartUpdatedEvent) MarshalJSON() ([]byte, error) {
	patch := e.Patch
	if patch == nil {
		patch = map[string]json.RawMessage{}
	}
	return json.Marshal(struct {
		Event     string                     `json:"event"`
		MessageID string                     `json:"message_id"`
		Index     int                        `json:"index"`
		Patch     map[string]json.RawMessage `json:"patch"`
	}{Event: EventPartUpdated, MessageID: e.MessageID, Index: e.Index, Patch: patch})
}

// MessageFinalizedEvent carries the authoritative final message.
type MessageFinalizedEvent struct {
	MessageID string
	Message   ChatMessage
}

func (MessageFinalizedEvent) EventName() string { return EventMessageFinalized }

func (e MessageFinalizedEvent) MarshalJSON() ([]byte, error) {
	return json.Marshal(struct {
		Event     string      `json:"event"`
		MessageID string      `json:"message_id"`
		Message   ChatMessage `json:"message"`
	}{Event: EventMessageFinalized, MessageID: e.MessageID, Message: e.Message})
}

// DecodeStreamEvent parses a wire event into its concrete type. Unknown event
// names return (nil, nil) so callers can drop them (forward-compat).
func DecodeStreamEvent(data []byte) (StreamEvent, error) {
	var head struct {
		Event string `json:"event"`
	}
	if err := json.Unmarshal(data, &head); err != nil {
		return nil, fmt.Errorf("spec: decode stream event: %w", err)
	}
	switch head.Event {
	case EventMessageStarted:
		var body struct {
			Message ChatMessage `json:"message"`
		}
		if err := json.Unmarshal(data, &body); err != nil {
			return nil, fmt.Errorf("spec: %s: %w", head.Event, err)
		}
		return MessageStartedEvent{Message: body.Message}, nil
	case EventPartAppended:
		var body struct {
			MessageID string      `json:"message_id"`
			Index     int         `json:"index"`
			Part      ContentPart `json:"part"`
		}
		if err := json.Unmarshal(data, &body); err != nil {
			return nil, fmt.Errorf("spec: %s: %w", head.Event, err)
		}
		return PartAppendedEvent{MessageID: body.MessageID, Index: body.Index, Part: body.Part}, nil
	case EventTokenDelta:
		var body struct {
			MessageID string `json:"message_id"`
			Index     int    `json:"index"`
			Text      string `json:"text"`
		}
		if err := json.Unmarshal(data, &body); err != nil {
			return nil, fmt.Errorf("spec: %s: %w", head.Event, err)
		}
		return TokenDeltaEvent{MessageID: body.MessageID, Index: body.Index, Text: body.Text}, nil
	case EventPartUpdated:
		var body struct {
			MessageID string                     `json:"message_id"`
			Index     int                        `json:"index"`
			Patch     map[string]json.RawMessage `json:"patch"`
		}
		if err := json.Unmarshal(data, &body); err != nil {
			return nil, fmt.Errorf("spec: %s: %w", head.Event, err)
		}
		return PartUpdatedEvent{MessageID: body.MessageID, Index: body.Index, Patch: body.Patch}, nil
	case EventMessageFinalized:
		var body struct {
			MessageID string      `json:"message_id"`
			Message   ChatMessage `json:"message"`
		}
		if err := json.Unmarshal(data, &body); err != nil {
			return nil, fmt.Errorf("spec: %s: %w", head.Event, err)
		}
		return MessageFinalizedEvent{MessageID: body.MessageID, Message: body.Message}, nil
	default:
		return nil, nil
	}
}

File: python/src/scitrera_messaging_spec/events.py
"""Streaming event vocabulary for the universal ChatMessage spec.

A small, ordered set of events that mutate an in-flight :class:`ChatMessage`
on the receiver. The final reconstructed message must equal the
``MessageFinalized.message`` payload exactly — i.e. the events are a
verbatim, sender-controlled view of how the canonical at-rest message was
assembled.

Usage pattern::

    state: dict[str, ChatMessage] = {}
    for event in stream:
        state = apply_event(state, event)
    # state[message_id] is now the final ChatMessage

A TypeScript sibling lives at ``typescript/src/applyEvent.ts`` and must
agree fixture-for-fixture with this reducer.
"""
from __future__ import annotations

from typing import Annotated, Any, Literal, Union

from pydantic import BaseModel, ConfigDict, Discriminator, Tag, TypeAdapter

from .schema import ChatMessage, ContentPart


class _EventBase(BaseModel):
    model_config = ConfigDict(extra="allow")


class MessageStarted(_EventBase):
    """Open a new in-flight message on the receiver."""

    event: Literal["message_started"] = "message_started"
    message: ChatMessage


class PartAppended(_EventBase):
    """Append a new content part at ``index`` (the resulting position)."""

    event: Literal["part_appended"] = "part_appended"
    message_id: str
    index: int
    part: ContentPart


class TokenDelta(_EventBase):
    """Append text to the ``text`` field of the part at ``index``.

    Hot path for streaming model output. The targeted part must be a
    ``text`` or ``reasoning`` part. Clients that don't want to special-case
    deltas may reconstruct via successive ``part_updated`` events on
    ``text`` instead.
    """

    event: Literal["token_delta"] = "token_delta"
    message_id: str
    index: int
    text: str


class PartUpdated(_EventBase):
    """Shallow-merge ``patch`` into the part at ``index``.

    Patch keys overwrite existing keys; ``None`` clears a key. Nested
    objects (e.g. ``meta``) are *not* deep-merged — re-emit the whole
    sub-object if you need to update it.
    """

    event: Literal["part_updated"] = "part_updated"
    message_id: str
    index: int
    patch: dict[str, Any]


class MessageFinalized(_EventBase):
    """Authoritative end-of-stream announcement.

    Carries the full canonical ``ChatMessage`` so clients that joined late,
    dropped events, or want to verify their reconstruction can sync to a
    known-good final state.
    """

    event: Literal["message_finalized"] = "message_finalized"
    message_id: str
    message: ChatMessage


StreamEvent = Annotated[
    Union[
        Annotated[MessageStarted, Tag("message_started")],
        Annotated[PartAppended, Tag("part_appended")],
        Annotated[TokenDelta, Tag("token_delta")],
        Annotated[PartUpdated, Tag("part_updated")],
        Annotated[MessageFinalized, Tag("message_finalized")],
    ],
    Discriminator(lambda v: v.get("event") if isinstance(v, dict) else getattr(v, "event", None)),
]

_event_adapter: TypeAdapter[StreamEvent] = TypeAdapter(StreamEvent)
_part_adapter: TypeAdapter[ContentPart] = TypeAdapter(ContentPart)


def parse_event(value: Any) -> Any:
    """Parse a JSON-shaped event dict (or model) into the right subclass."""
    return _event_adapter.validate_python(value)


def apply_event(
    messages: dict[str, ChatMessage],
    event: Any,
) -> dict[str, ChatMessage]:
    """Pure reducer: apply ``event`` to a ``{message_id: ChatMessage}`` map.

    Returns a new dict (does not mutate the input).
    """
    parsed = event if isinstance(event, _EventBase) else parse_event(event)

    if isinstance(parsed, MessageStarted):
        return {**messages, parsed.message.id: parsed.message.model_copy(deep=True)}

    if isinstance(parsed, PartAppended):
        msg = messages.get(parsed.message_id)
        if msg is None:
            return messages
        new_content = list(msg.content)
        part = (
            parsed.part
            if isinstance(parsed.part, BaseModel)
            else _part_adapter.validate_python(parsed.part)
        )
        idx = max(0, min(parsed.index, len(new_content)))
        new_content.insert(idx, part)
        return {**messages, msg.id: msg.model_copy(update={"content": new_content})}

    if isinstance(parsed, TokenDelta):
        msg = messages.get(parsed.message_id)
        if msg is None or not (0 <= parsed.index < len(msg.content)):
            return messages
        part = msg.content[parsed.index]
        cur = getattr(part, "text", None)
        if not isinstance(cur, str):
            return messages
        new_part = part.model_copy(update={"text": cur + parsed.text})
        new_content = list(msg.content)
        new_content[parsed.index] = new_part
        return {**messages, msg.id: msg.model_copy(update={"content": new_content})}

    if isinstance(parsed, PartUpdated):
        msg = messages.get(parsed.message_id)
        if msg is None or not (0 <= parsed.index < len(msg.content)):
            return messages
        part = msg.content[parsed.index]
        merged = {**part.model_dump(exclude_none=False), **parsed.patch}
        new_part = _part_adapter.validate_python(merged)
        new_content = list(msg.content)
        new_content[parsed.index] = new_part
        return {**messages, msg.id: msg.model_copy(update={"content": new_content})}

    if isinstance(parsed, MessageFinalized):
        return {**messages, parsed.message_id: parsed.message.model_copy(deep=True)}

    return messages


__all__ = [
    "MessageStarted",
    "PartAppended",
    "TokenDelta",
    "PartUpdated",
    "MessageFinalized",
    "StreamEvent",
    "parse_event",
    "apply_event",
]
